### DISCUSSION TITLE
The problem of running slower and slower

### DISCUSSION THREAD

**Discussion by rebornfz** [libvips/libvips#2254] [Q&A] [answered]
Hello, I implemented an image rendering service using libvips. But rendering takes longer day by day. The average time is getting longer and longer.
Both memory usage and cpu usage are in the normal state. 
My cpu is Intel(R) Xeon(R) Platinum 8255C CPU @ 2.50GHz.
Even when it takes a long time, cpu usage is not high.
I don't set the vips_concurrency_set (). Maybe I need to limit the number of worker threads?
What should I pay attention to when the program is running for a long time?

**Comment by jcupitt** [comment]
Hi @rebornfz, that sounds very strange.

What language and libvips binding are you using? What platform are you running on? Can you share your code? Could you perhaps have a reference leak?

You can set the `VIPS_LEAK` environment variable to make libvips track references and report leaks. For example:

```
$ VIPS_LEAK=1 vips copy k2.jpg k2.png
memory: high-water mark 28.68 MB
```

I would set the variable, start your service, run for 10 minutes under load, then exit normally (don't abort, just call `exit(0)`) and see if it prints anything.

**Comment by rebornfz** [comment]
@jcupitt 
I am using c++ and the version of vips is vips-8.10.6. 
The platform is centos7. 
Sorry, for some reasons, I cannot provide all the code.
I will try to set the VIPS_LEAK.

**Comment by rebornfz** [comment]
@jcupitt  I found the problem. I composite a array of images.  And when an image only has two bands(generally has 4 bands), I do like this:
```
image_arr[i] = vips::VImage::pngload_buffer(image_arr[i].pngsave_buffer());
``` 

```
        for(int i = 0; i < image_arr.size(); i++)
        {
            image_arr[i] = image_arr[i].copy(vips::VImage::option ()->
                set("interpretation", VIPS_INTERPRETATION_sRGB));
            //std::cout << image_arr[i].bands() << std::endl;
            if(image_arr[i].bands() == 2)
            {
                LOG(INFO) << "bands = 2" << std::endl;
                image_arr[i] = vips::VImage::pngload_buffer(image_arr[i].pngsave_buffer());
            }
        }
        image_output = vips::VImage::composite(image_arr, mode, 
        vips::VImage::option ()->
        set("x", arrx)->
        set("y", arry));
```
And this lead a leak. How can I modify this code?

1 VipsArea alive
	0x3505e00 count = 1, bytes = 13209
memory: high-water mark 92.31 MB

**Reply by jcupitt** [reply]
Ah that will be very slow, yes. Could you post a larger piece of code? I'm not sure I understand what you are doing.

**Reply by jcupitt** [reply]
The leak is the result of `pngsave_buffer` not being freed. But this is not a good way to fix the number of image bands. 

`composite` should convert mono to RGBA for you automatically. For example, here's the standard PNG transparency demo image:

![PNG_transparency_demonstration](https://user-images.githubusercontent.com/580843/118272578-b8117280-b4ba-11eb-9d8e-8f1728d844f6.png)

I can make a mono version like this:

```
$ vips colourspace PNG_transparency_demonstration.png png-mono.png b-w
```

Then composite the colour and mono versions on to a base JPG like this:

```
$ vips composite "Gugg_coloured.jpg png-mono.png PNG_transparency_demonstration.png" x.png 12 --x "0 300" -y "0 300"
```

(12 is `over`) to make this:

![x](https://user-images.githubusercontent.com/580843/118272757-fe66d180-b4ba-11eb-8f7d-c749723660e6.png)

`composite` moves all input images to the compositing space (srgb by default) before it starts, so it's fine to mix mono and coloured images together, and you shouldn't need to test for bands.

Perhaps your two band images have not been tagged as being mono?

**Reply by rebornfz** [reply]
@jcupitt  How can I tag the two bands images as being mono?
If I don't do this
```
image_arr[i] = vips::VImage::pngload_buffer(image_arr[i].pngsave_buffer());
```
The composite will go wrong, when an image has only two band.

**Comment by jcupitt** [comment, marked as answer]
You can set the interpretation to mono in the same way you set srgb:

```C++
x = x.copy(vips::VImage::option ()->set("interpretation", VIPS_INTERPRETATION_B_W));
```

(`b_w` is black and white, for historical reasons). I'd check the code that made the GA (grey + alpha) image and add the interpretation hint there.
